### ISSUE TITLE
[EPCO-19] Nearest decision boundary points to a query point can be found

### ISSUE THREAD

**Issue by hatfield-c** [hatfield-c/epco#19]
Given a query point q, k uniformly distributed directions around the point will end with the point of intersection to nearest decision boundary along each given direction, up to a certain distance. If a decision boundary is not encountered within the distance, then that point at maximum distance in valid space is returned.

**Comment by hatfield-c** [comment]
Copied from EPCO-#6:

We might be able to use Newton's Method of 2nd order optimization to quickly find linear translations through the p-field input space that are valid.

The issue with straight gradient descent in the p-field is that choosing a good step size becomes very difficult when trying to deal with universal conditions. The potential for overstepping, or choosing a linear path that teleports through an invalid region will prove quite problematic, especially for problems that involve continuous dynamic path planning.

A 2nd order method like Newton's Method might in theory allow us to examine potential linear paths by guiding the proposed path endpoint straight to a decision boundary in a direction of interest. The dot product of the gradient at a point with a particular direction gives us the rate of change for that direction. Typically one can just choose the gradient during descent optimization, because it gives the direction of steepest descent. However I am hopeful that either all or most of the rules and conditions chosen for descent optimization will also work with non-steepest direction.

That is, we can choose k uniformly distributed directions from a particular point q, and use Newton's Method to converge the endpoints lines in these directions to the first decision boundary they intersect.
- If a cilia line detects that it is going uphill, it will revert to hill climbing behavior, and then it will go back to descent behavior once over the hill

**Comment by hatfield-c** [comment]
I'm still concerned about the "smoothness"of the function. Perhaps taking an average of points over an area will give smoother movement, but there is concern too for potential loss of accuracy.

**Comment by hatfield-c** [comment]
Maybe grow the cilia out a certain distance linearly, and then allow natural exploration?

**Comment by hatfield-c** [comment]
Newton's method has well known issues if you don't begin with an initial "good guess", which presents a problem for the current application. Furthermore Newton's method works by computing the x position which is the x-intercept of the line/plane given by the gradient with the x-axis. This approximation gets closer and closer to the "root" (where the function is 0). However, if the current position has a small gradient [-0.2453, 0.6653] then the stepsize will balloon out of control. Thus the method is generally considered to be best used when you are already able to somewhat start near your goal.

There might still be some use for a Newton-like approach, especially given its widespread use for hill climbing.

I have learned however that Newton's method is a subclass of optimization methods called Householder's method:
- https://en.wikipedia.org/wiki/Householder%27s_method

There might prove to be better methods for what I'm trying to do.

**Comment by hatfield-c** [comment]
We are generally only interested in finding roots that are within a certain distance. Maybe this fact can be exploited.

**Comment by hatfield-c** [comment]
On Householder's method:
"This method, like Newton's method, has poor convergence properties near any point where the derivative equals 0."
- https://mathworld.wolfram.com/HouseholdersMethod.html

**Comment by hatfield-c** [comment]
So here is an obvious point that seems to have escaped me thus far: this is a ReLU network, hence any one particular point has a hessian of 0, as all points of the function are localized planar approximations, and the gradients of planes don't change.

Thus all second order methods are not applicable, including Newton's method.

A smoothing strategy via averaging would help mitigate the effect, but only partially.

**Comment by hatfield-c** [comment]
The most effective mitigation measure I can think of for the "choppiness" of the distance field would be to train the network such that points closer to the decision boundary have a lower distance than points farther from the boundary. 

I will make a new issue for this.

**Comment by hatfield-c** [comment]
It is presently possible to navigate from invalid space to positive space with the newly rebased VNN code. However, locating the decision boundary has still not yet been accomplished.

**Comment by hatfield-c** [comment]
The VNN (Voxel Neural Network) approach gives the exact distance to all decision boundaries for each side of all voxels, so long as the voxel side weights are 1.

This issue is this complete and will now be closed.
